### PR TITLE
Various improvements for v0.1

### PR DESCRIPTION
**Pull request by JohannesMeierSE**
Beyond improving comments, refactorings, polishing and so on, these notable improvements are done in this PR:
- functions and operators can be annotated with validations now (review: should we support that for all our types?)
- Typir module:
    - reorganized the grouping of services
    - don't merge Langium+Typir services, but provide the Typir services via an "typir"-property/service within the Langium services
    - override only selected services of a group
    - extracted missing service interface
    - extracted logic to resolve TypeSelectors into its own service to make the implementation exchangable
    - the module now looks into the kind registry, whether the kind is already initialized, before a new one is created (this partially fulfills #44)
- removed dependencies to 'langium/lsp' from Typir-Langium in order to use it also in non-LSP projects
- switched to Langium v3.3, simplified API, removed duplicated code for a fix which is part of Langium 3.3 now
- added missing READMEs, ChangeLog, added content, added EclipseCon slides
- associate domain elements with Typir types
- implemented overloaded functions and methods in LOX by using Typir already during Linking!
- reduced code for the "Tiny Typir" example a bit

Hints for the review
- Sorry for the huge number of changes!
- Therefore, I suggest to review commit by commit.
- Feel free to ask for a joint review session!

### CODE REVIEW

File: examples/lox/test/lox-type-checking-functions.test.ts [hunk 1]
```diff
@@ -4,43 +4,94 @@
...
+        const method2 = call2Node.element?.ref;
+        assertTrue(isFunctionDeclaration(method2));
+        expect(method2.returnType.primitive).toBe('boolean');
+        // check type inference
+        const call2Type = loxServices.typir.Inference.inferType(call2Node);
+        expect(isType(call2Type)).toBeTruthy();
+        assertType(call2Type, isPrimitiveType);
+        expect(call2Type.getName()).toBe('boolean');
+    });
+
```
Comment (insafuhrmann) [inline examples/lox/test/lox-type-checking-functions.test.ts:96]: Thanks, the new testcases are helpful for documentation purposes as well.

File: examples/lox/src/language/lox-type-checking.ts [hunk 2]
```diff
@@ -63,7 +63,7 @@ export class LoxTypeCreator extends AbstractLangiumTypeCreator {
         for (const operator of ['-', '*', '/']) {
             this.typir.factory.Operators.createBinary({ name: operator, signature: { left: typeNumber, right: typeNumber, return: typeNumber }, inferenceRule: binaryInferenceRule });
         }
-        this.typir.factory.Operators.createBinary({ name: '+', signature: [
+        this.typir.factory.Operators.createBinary({ name: '+', signatures: [
```
Comment (insafuhrmann) [inline examples/lox/src/language/lox-type-checking.ts:66]: Nice, thanks :)

**Review by insafuhrmann — APPROVED** [review]
Thanks for this big lot of improvements @JohannesMeierSE. They look fine to me, I just have a few detail remarks.

File: examples/lox/src/language/type-system/lox-type-checking.ts [hunk 3]
```diff
@@ -45,7 +45,7 @@ export class LoxTypeCreator extends AbstractLangiumTypeCreator {
                 (node: unknown) => isReturnStatement(node) && node.value === undefined
             ] });
         const typeNil = this.typir.factory.primitives.create({ primitiveName: 'nil',
-            inferenceRules: isNilLiteral }); // From "Crafting Interpreters" no value, like null in other languages. Uninitialised variables default to nil. When the execution reaches the end of the block of a function body without hitting a return, nil is implicitly returned.
+            inferenceRules: isNilLiteral }); // From "Crafting Interpreters" no value, like null in other languages. TODO Uninitialised variables default to nil. When the execution reaches the end of the block of a function body without hitting a return, nil is implicitly returned.
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply examples/lox/src/language/type-system/lox-type-checking.ts]: Already fixed: I slightly changed LOX here to "'nil' is only assignable to variables with a class as type in the LOX implementation here"

File: examples/lox/src/language/type-system/lox-type-checking.ts [hunk 4]
```diff
@@ -116,28 +116,23 @@ export class LoxTypeCreator extends AbstractLangiumTypeCreator {
...
             // ... variable declarations
             if (isVariableDeclaration(domainElement)) {
                 if (domainElement.type) {
-                    // the user declared this variable with a type
-                    return domainElement.type;
+                    return domainElement.type; // the user declared this variable with a type
                 } else if (domainElement.value) {
-                    // the didn't declared a type for this variable => do type inference of the assigned value instead!
-                    return domainElement.value;
+                    return domainElement.value; // the user didn't declared a type for this variable => do type inference of the assigned value instead!
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply examples/lox/src/language/type-system/lox-type-checking.ts:134]: fixed

File: packages/typir/src/typir.ts [hunk 5]
```diff
@@ -23,22 +23,20 @@ import { DefaultValidationCollector, DefaultValidationConstraints, ValidationCol
...
  */
 
-/** Open design questions TODO
- * - use graphology for the TypeGraph?
- * - Where should inference rules be stored? only in the central service? in types? in kinds?
- * - Type is generic VS there are specific types like FunctionType (extends Type)?? functionType.kind.getOutput(functionKind) + isFunctionKind() feels bad! vs functionType.getOutput() + isFunctionType()
- * - realize "unknown" as a generic "<T = unknown>" for whole Typir? for the Langium binding T would be AstNode!
- * - Is it easy to use two different Typir instances side-by-side within the same application?
+/** Some open design questions for future releases TODO
+ * - Replace "unknown" as a generic "<T = unknown>" for whole Typir? For Typir-Langium T would be AstNode!
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply packages/typir/src/typir.ts:39]: I created #49

File: CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  - Sub-typing
+  - Validation
+  - Caching
+- [Predefined types](/packages/typir/src/kinds/) to reuse:
+  - Primitives
+  - Functions
+  - Classes (nominally typed)
+  - Top, bottom
+  - (some more are under development)
+  - Operators (which are mapped to Functions)
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply CHANGELOG.md:27]: Thanks! I mentioned it here and in the main README.md

File: packages/typir/src/graph/type-node.ts [hunk 7]
```diff
@@ -63,7 +63,11 @@ export abstract class Type {
     protected readonly edgesIncoming: Map<string, TypeEdge[]> = new Map();
     protected readonly edgesOutgoing: Map<string, TypeEdge[]> = new Map();
 
-    /** An element from the domain might be associated with this type, e.g. the declaration node in the AST. */
+    /**
+     * An element from the domain might be associated with this type, e.g. the declaration node in the AST.
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply packages/typir/src/graph/type-node.ts:67]: fixed, additionally I reworked the description

File: examples/lox/src/language/type-system/lox-type-checking.ts [hunk 8]
```diff
@@ -83,10 +83,22 @@ export class LoxTypeCreator extends AbstractLangiumTypeCreator {
 
         // ==, != for all data types (the warning for different types is realized below)
         for (const operator of ['==', '!=']) {
-            this.typir.factory.operators.createBinary({ name: operator, signature: { left: typeAny, right: typeAny, return: typeBool }, inferenceRule: binaryInferenceRule });
+            this.typir.factory.operators.createBinary({ name: operator, signature: { left: typeAny, right: typeAny, return: typeBool }, inferenceRule: binaryInferenceRule,
+                // show a warning to the user, if something like "3 == false" is compared, since different types already indicate, that the IF condition will be evaluated to false
+                validationRule: (node, _operatorName, _operatorType, typir) => typir.validation.constraints.ensureNodeIsEquals(node.left, node.right, (actual, expected) => <ValidationMessageDetails>{
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply examples/lox/src/language/type-system/lox-type-checking.ts]: I added a longer comment about this.

File: examples/lox/src/language/lox-linker.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                            const parameterTypes = called.parameters.map(p => this.typir.Inference.inferType(p)).filter(isType);
+                            if (parameterTypes.length === called.parameters.length) { // for all parameters, a type is inferred
+                                if (argumentTypes.every((arg, index) => this.typir.Assignability.isAssignable(arg, parameterTypes[index]))) {
+                                    return calledDescription;
+                                }
+                            }
+                        }
+                    }
+                }
+                // no matching method is found, return the first found method => linking works + validation issues regarding the wrong parameter values can be shown!
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply examples/lox/src/language/lox-linker.ts:50]: You are right, I created #50 to fix that for v0.2.

File: examples/lox/src/language/lox-linker.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                return calledDescriptions[0];
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply examples/lox/src/language/lox-linker.ts:51]: You are right, I created #50 to fix that for v0.2.

File: packages/typir/src/services/validation.ts [hunk 11]
```diff
@@ -142,41 +142,41 @@ export class DefaultValidationConstraints implements ValidationConstraints {
 }
 
 
-export interface ValidationCollector {
-    validateBefore(domainRoot: unknown): ValidationProblem[];
-    validate(domainElement: unknown): ValidationProblem[];
-    validateAfter(domainRoot: unknown): ValidationProblem[];
+export interface ValidationCollector<ElementType = unknown, RootType = ElementType> {
+    validateBefore(domainRoot: RootType): ValidationProblem[];
+    validate(domainElement: ElementType): ValidationProblem[];
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply packages/typir/src/services/validation.ts:147]: I will do all these renamings in a follow-up PR.

File: examples/lox/src/language/lox-scope.ts [hunk 12]
```diff
@@ -32,9 +38,10 @@ export class LoxScopeProvider extends DefaultScopeProvider {
             if (!previous) {
                 return super.getScope(context);
             }
-            const previousType = inferType(previous, new Map());
+            // use Typir to identify the ClassType of the current expression (including variables, fields of nested classes, ...)
+            const previousType = this.typir.Inference.inferType(previous);
             if (isClassType(previousType)) {
-                return this.scopeClassMembers(previousType.literal);
+                return this.scopeClassMembers(previousType.associatedDomainElement as Class); // the Class was associated with this ClassType during its creation
```
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply examples/lox/src/language/lox-scope.ts:44]: I will do all these renamings in a follow-up PR.

**Review by insafuhrmann — APPROVED** [review]
Thanks for the improvements, looks fine to me!

**Pull request merged**